packages/plugins/response-cache/src/index.ts

```diff
@@ -129,61 +126,69 @@ export function useResponseCache(options: UseResponseCacheParameter): Plugin {
         }),
       );
     },
-    async onRequest({ request, fetchAPI, endResponse }) {
-      if (enabled(request)) {
+    onRequest({ request, serverContext, fetchAPI, endResponse }): PromiseOrValue<void> {
+      if (enabled(request, serverContext as TContext)) {
         const operationId = request.headers.get('If-None-Match');
         if (operationId) {
-          const cachedResponse = await cache.get(operationId);
-          if (cachedResponse) {
-            const lastModifiedFromClient = request.headers.get('If-Modified-Since');
-            const lastModifiedFromCache =
-              cachedResponse.extensions?.http?.headers?.['Last-Modified'];
-            if (
-              // This should be in the extensions already but we check it here to make sure
-              lastModifiedFromCache != null &&
-              // If the client doesn't send If-Modified-Since header, we assume the cache is valid
-              (lastModifiedFromClient == null ||
-                new Date(lastModifiedFromClient).getTime() >=
-                  new Date(lastModifiedFromCache).getTime())
-            ) {
-              const okResponse = new fetchAPI.Response(null, {
-                status: 304,
-                headers: {
-                  ETag: operationId,
-                },
-              });
-              endResponse(okResponse);
+          return mapMaybePromise(cache.get(operationId), cachedResponse => {
+            if (cachedResponse) {
+              const lastModifiedFromClient = request.headers.get('If-Modified-Since');
+              const lastModifiedFromCache =
+                cachedResponse.extensions?.http?.headers?.['Last-Modified'];
+              if (
+                // This should be in the extensions already but we check it here to make sure
+                lastModifiedFromCache != null &&
+                // If the client doesn't send If-Modified-Since header, we assume the cache is valid
+                (lastModifiedFromClient == null ||
+                  new Date(lastModifiedFromClient).getTime() >=
+                    new Date(lastModifiedFromCache).getTime())
+              ) {
+                const okResponse = new fetchAPI.Response(null, {
+                  status: 304,
+                  headers: {
+                    ETag: operationId,
+                  },
+                });
+                endResponse(okResponse);
+              }
             }
-          }
+          }) as Promise<void>;
```

Comment:
This is necessary because the onRequest hook in whatwg expects a promise? I dont like this assertion. I'd prefer to have it fixed in whatwg, and use @ts-expect-error here in the meantime.

Reply:
Reverted the change anyways.